### DISCUSSION TITLE
Manual certificate verification seems to be regressed in 7.80.0

### DISCUSSION THREAD

**Discussion by biao-billy** [curl/curl#8576] [Q&A]
Hi all, hi @bagder.

My code performs manual server certificate verification (via SSL_CTX_set_verify(..., SSL_VERIFY_PEER, NULL) and SSL_CTX_set_cert_verify_callback(..., _ssl_app_verify_callback, _this) ).

I upgraded to 7.80.0 my X509_verify_cert call started to fail with "unable to get local issuer certificate (20)". The code functions normally with 7.79.0.

Looking at the change log I noticed a change "openssl: if verifypeer is not requested, skip the CA loading". I wonder if this change caused a behavior change that led to error 20.

If it is, can you please suggest a workaround?

Thanks,
Billy

**Comment by bagder** [comment]
Then it would help to get that confirmed. Did 83393b1a3613 cause this regression for you? Is your callback thus assuming that libcurl loaded the CA certs even though you didn't ask it to verify them?

**Reply by biao-billy** [reply]
Hi @bagder, let me patch 7.80 with that commit reverted and get back to you.

Yes I just call curl_easy_setopt(curl, CURLOPT_CAINFO, ...) currently. Looking at the commit, looks like adding a SSL_CTX_load_verify_locations call in the callback should just fix my issue - albeit a bit redundant as I've told curl about the CA file location.

**Reply by biao-billy** [reply]
Hi @bagder. I can confirm reverting 83393b1 fixed my problem. I can also confirm I could work around the problem by loading the CA file via SSL_CTX_load_verify_locations from my CURLOPT_SSL_CTX_FUNCTION.

IMO you may want to consider scoping the new logic only when the app does not call SSL_CTX_set_cert_verify_callback.

I'll let you decide whether this is a bug? Or a breaking change.

Thanks, Billy.
